build/cmake/FindSASL2.cmake

```diff
@@ -1,54 +1,74 @@
-        sasl_client_done
-        sasl/sasl.h
-        HAVE_SASL_CLIENT_DONE)
-    cmake_pop_check_state()
-
-    if (HAVE_SASL_CLIENT_DONE)
-        set (MONGOC_HAVE_SASL_CLIENT_DONE 1)
-    else ()
-        set (MONGOC_HAVE_SASL_CLIENT_DONE 0)
-    endif ()
-else ()
-    if (ENABLE_SASL STREQUAL AUTO)
-        set (SASL_FOUND 0)
-        set (SASL_INCLUDE_DIRS "")
-        set (SASL_LIBRARIES "")
-        set (MONGOC_HAVE_SASL_CLIENT_DONE 0)
-    else ()
-        message (FATAL_ERROR "  SASL not found")
-    endif ()
-endif ()
+#[[
+
+Searches for a Cyrus "libsasl2" library available on the system using
+pkg-config. The libsasl2.pc file must be available for pkg-config!
+
+Upon success, Defines an imported target `SASL2::SASL2` that can be linked into
+other targts.
+
+]]
+
+if(SASL2_FIND_COMPONENTS)
+    message(FATAL_ERROR "This find_package(SASL2) does not support package components (Got “${SASL2_FIND_COMPONENTS}”)")
+endif()
+set(_modspec "libsasl2")
+if(DEFINED SASL2_FIND_VERSION)
+    set(_modspec "libsasl2>=${SASL2_FIND_VERSION}")
+endif()
+
+if(TARGET SASL2::SASL2)
+    message(DEBUG "SASL2 ${SASL2_VERSION} was already found")
+    # There is already a libsasl2 imported
+    if(NOT DEFINED SASL2_FIND_VERISON OR SASL2_FIND_VERSION VERSION_LESS_EQUAL SASL2_VERSION)
+        # Okay: We are satisfied by the version that has already been found
+        set(SASL2_FOUND TRUE)
+        return()
+    endif()
+    # Eh: The importer has requested a version of libsasl2 that is greater than the
+    # version of libsasl2 that we previously imported. The pkg-config invocation
+    # will likely fail, but we'll let it say that for itself.
+    message(DEBUG "Need to find SASL2 again: Version requirement increased ${SASL2_FIND_VERSION} > ${SASL2_VERSION}")
+endif()
+
+# Upon early return, tell the caller that we don't have it:
+set(SASL2_FOUND FALSE)
+
+# Forward the QUIET+REQUIRED args to pkg-config
+set(_required)
+set(_quiet)
+if(SASL2_FIND_QUIETLY)
+    set(_quiet QUIET)
+endif()
+if(SASL2_FIND_REQUIRED)
+    set(_required REQUIRED)
+endif()
+
+# We use pkg-config to find libsasl2.
+find_package(PkgConfig ${_required} ${_quiet})
+if(NOT PkgConfig_FOUND)
+    # The find_package(PkgConfig) will have generated errors for us if we were REQUIRED,
+    # so we can just return without setting any message
+    set(SASL2_NOT_FOUND_MESSAGE [[No pkg-config executable was found]])
+    return()
+endif()
+
+# Now ask pkg-config to find libsasl2 for us
+message(DEBUG "Using pkg-config to search for module “${_modspec}”")
+pkg_check_modules(libsasl2 ${_required} ${_quiet} IMPORTED_TARGET "${_modspec}" GLOBAL)
+# Note: pkg_check_modules() does support static libs, but we don't use them here,
+# and static libsasl2 is uncommon.
+if(NOT libsasl2_FOUND)
+    # The pkg_check_modules() will have issued the error for us if we were REQUIRED.
+    set(SASL2_NOT_FOUND_MESSAGE [[pkg-config failed to find a libsasl2 (Cyrus) dynamic library package]])
+    return()
+endif()
```

Comment:
Before these changes: libsasl2 is located on macOS. Here is the output of cmake:

```
-- Searching for sasl/sasl.h
--   Found in /Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk/usr/include
-- Searching for libsasl2
--   Found /Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk/usr/lib/libsasl2.tbd
```

After these changes: libsasl2 is not found.

The install of libsasl2 on macOS does not appear to include pkgconfig `.pc` files:

``` 
% find /Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk/ -name "*sasl*.pc"
(no results)
```

The `cyrus-sasl` Homebrew package includes pkg-config files, but will not install to default locations. `brew install cyrus-sasl` outputs these messages:

```
cyrus-sasl is keg-only, which means it was not symlinked into /opt/homebrew,
because macOS already provides this software and installing another version in
parallel can cause all kinds of trouble.
```

And:

```
For pkg-config to find cyrus-sasl you may need to set:
  export PKG_CONFIG_PATH="/opt/homebrew/opt/cyrus-sasl/lib/pkgconfig"
```

Can the prior behavior (search for header + library) be added as a fallback to prevent breaking macOS configurations?

Reply:
I think I will revert to the prior behavior, as the pkg-config impl is fragile in several ways.

Reply:
Confirmed building with b2c72b662cfeed8677d72d1b9a9d1013a21b8641 is able to locate libsasl2 on my macOS.